internal/app/machined/pkg/controllers/block/devices.go

```diff
@@ -0,0 +1,193 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package block
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+
+	"github.com/cosi-project/runtime/pkg/controller"
+	"github.com/cosi-project/runtime/pkg/safe"
+	"github.com/cosi-project/runtime/pkg/state"
+	"go.uber.org/zap"
+
+	"github.com/siderolabs/talos/internal/app/machined/pkg/controllers/block/internal/kobject"
+	"github.com/siderolabs/talos/internal/app/machined/pkg/controllers/block/internal/sysblock"
+	machineruntime "github.com/siderolabs/talos/internal/app/machined/pkg/runtime"
+	"github.com/siderolabs/talos/pkg/machinery/resources/block"
+)
+
+// DevicesController provides a view of available block devices with information about pending updates.
+type DevicesController struct {
+	V1Alpha1Mode machineruntime.Mode
+}
+
+// Name implements controller.Controller interface.
+func (ctrl *DevicesController) Name() string {
+	return "block.DevicesController"
+}
+
+// Inputs implements controller.Controller interface.
+func (ctrl *DevicesController) Inputs() []controller.Input {
+	return nil
+}
+
+// Outputs implements controller.Controller interface.
+func (ctrl *DevicesController) Outputs() []controller.Output {
+	return []controller.Output{
+		{
+			Type: block.DeviceType,
+			Kind: controller.OutputExclusive,
+		},
+	}
+}
+
+// Run implements controller.Controller interface.
+//
+//nolint:gocyclo
+func (ctrl *DevicesController) Run(ctx context.Context, r controller.Runtime, logger *zap.Logger) error {
+	// in container mode, no devices
+	if ctrl.V1Alpha1Mode == machineruntime.ModeContainer {
+		return nil
+	}
+
+	// start the watcher first
+	watcher, err := kobject.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("failed to create kobject watcher: %w", err)
+	}
+
+	defer watcher.Close() //nolint:errcheck
+
+	watchCh := watcher.Run(logger)
+
+	// reconcile the initial list of devices while the watcher is running
+	select {
+	case <-ctx.Done():
+		return nil
+	case <-r.EventCh():
+	}
+
+	if err = ctrl.resync(ctx, r, logger); err != nil {
+		return fmt.Errorf("failed to resync: %w", err)
+	}
+
+	for {
+		select {
+		case ev := <-watchCh:
+			if ev.Subsystem != "block" {
+				continue
+			}
+
+			ev.DevicePath = filepath.Join("/sys", ev.DevicePath)
+
+			if err = ctrl.processEvent(ctx, r, logger, ev); err != nil {
+				return err
+			}
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
+
+func (ctrl *DevicesController) resync(ctx context.Context, r controller.Runtime, logger *zap.Logger) error {
+	events, err := sysblock.Walk("/sys/block")
+	if err != nil {
+		return fmt.Errorf("failed to walk /sys/block: %w", err)
+	}
+
+	touchedIDs := make(map[string]struct{}, len(events))
```

Comment:
`touchedIDs` supposed to be scoped to `resync` lifetime?

Reply:
yes